pkg/controller/controller_instance.go

```diff
@@ -169,7 +168,7 @@ func (c *controller) reconcileServiceInstanceDelete(instance *v1alpha1.ServiceIn
 	//
 	// TODO: the above logic changes slightly once we handle orphan
 	// mitigation.
-	if !instance.Status.AsyncOpInProgress && instance.Status.Checksum == nil {
+	if !instance.Status.AsyncOpInProgress && instance.Status.ReconciledGeneration == 0 {
```

Comment:
Why does it being set to zero matter?

Reply:
If ReconciledGeneration is 0, then this is a ServiceInstance that has not yet been provisioned. Because it has not been provisioned, there is no need to send a DeprovisionRequest to the broker.

Reply:
hmmm, what if it was provisioned but we didn't get the response. I'm pretty sure we'll be in orphan mitigation stuff, but when we delete the object does it stop orphan mitigation or let it finish?  Also, who added the finalizerToken and do we need to remove it? Just want to make sure we're not leaving stuff around.

Reply:
The finalizer token is added by the instance REST strategy in `PrepareForCreate`. As far as I can glean, deleting the finalizer indicates that service-catalog no longer has any use for the resource. If the controller does not delete the finalizer, than the resource will remain longer than it should as the registry will think that service-catalog is still using the resource.

Reply:
In the existing code, the checksum is only set when the status of the resource is updated with a Ready condition of true. That happens only when the Controller receives a success response from the Broker. So, even under the existing code, there is the potential gap when the Broker accepted the ProvisionRequest but the Controller did not receive the success response. I agree that that is a possibility and deserves an issue. Perhaps that is something that will be covered by orphan mitigation. I do not think that it is within the scope of this issue, however.

Reply:
> If the controller does not delete the finalizer, than the resource will remain longer than it should as the registry will think that service-catalog is still using the resource

But that's kind of my point. It seems to me that regardless of the ReconciledGeneration value we'll always want to delete the finalizer, no?

re: Ready condition - the ReconciledGeneration will get set even if talking to the broker fails.

Reply:
Eventually, yes, we always want to delete the finalizer. However, we may not want to delete the finalizer yet. For example, if there is a communication error when sending the DeprovisionRequest to the Broker, then the Controller will requeue the instance and attempt to perform the deprovision again later. In that case, we do not want to delete the finalizer yet.

Reply:
re: Ready condition - It is not my understanding that the ReconciledGeneration will get set if there was a communication error talking to the Broker. The ReconciledGeneration is only set when (1) the Broker responded with a success, (2) the Broker responded with a failure indicating that the request will not succeed, or (3) the retry duration has expired.

Reply:
but if we don't want to delete the finalizer because more stuff is going on then shouldn't the check be for Generation != ReconciledGeneration, not ReconciledGeneration !=0 ?  It just seems to me that the zero vs non-zero value here isn't the important thing, but perhaps I'm still just confused :-)

Reply:
Once DeletionTimestamp is not nil, the Controller does not care at all about the comparison between Generation and ReconciledGeneration. The comparison between the two is only relevant in the context of Provision and UpdateInstance. We are not postponing the deletion of the finalizer because there are more Provision or UpdateInstance tasks to do. We are postponing the deletion of the finalizer because there is a Deprovision task to do. In the case where ReconciledGeneration is zero, there is no Deprovision task to do because the Provision task was never completed.

Reply:
> The finalizer token is added by the instance REST strategy in PrepareForCreate

What if users add their own finalizers though? It's common for Kubernetes core, don't we support this in Service Catalog? Probably the more correct would be removing only "our" finalizers.

> In the case where ReconciledGeneration is zero, there is no Deprovision task to do because the Provision task was never completed.

There could be a cleanup (orphan mitigation) stuff to do (as @duglin already pointed out), and I would at least leave a more clear TODO comment for this there, if it is out of scope of this PR.

P.S. I understand that these issues are already present in `master`, and you're just replacing checksum with generation. So these questions are out of scope of this PR, and not blocking it from being merged, agreed.

Reply:
> What if users add their own finalizers though? It's common for Kubernetes core, don't we support this in Service Catalog? Probably the more correct would be removing only "our" finalizers.

The Controller is only removing the finalizer that Service Catalog added. It is leaving the other finalizers untouched on the resource.